ClientUserCodeDeploymentTest.testWithMultipleNodes_clientReconnectsToNewNode

```
com.hazelcast.nio.serialization.HazelcastSerializationException: java.lang.ClassNotFoundException: Failed to load class usercodedeployment.IncrementingEntryProcessor from other members.
	at com.hazelcast.internal.serialization.impl.JavaDefaultSerializers$JavaSerializer.read(JavaDefaultSerializers.java:83)
	at com.hazelcast.internal.serialization.impl.JavaDefaultSerializers$JavaSerializer.read(JavaDefaultSerializers.java:72)
	at com.hazelcast.internal.serialization.impl.StreamSerializerAdapter.read(StreamSerializerAdapter.java:48)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toObject(AbstractSerializationService.java:191)
	at com.hazelcast.client.impl.protocol.task.map.MapExecuteOnAllKeysMessageTask.createOperationFactory(MapExecuteOnAllKeysMessageTask.java:47)
	at com.hazelcast.client.impl.protocol.task.AbstractAllPartitionsMessageTask.processMessage(AbstractAllPartitionsMessageTask.java:38)
	at com.hazelcast.client.impl.protocol.task.AbstractMessageTask.initializeAndProcessMessage(AbstractMessageTask.java:130)
	at com.hazelcast.client.impl.protocol.task.AbstractMessageTask.run(AbstractMessageTask.java:110)
	at java.util.concurrent.ThreadPoolExecutor$Worker.runTask(ThreadPoolExecutor.java:895)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:918)
	at java.lang.Thread.run(Thread.java:662)
	at com.hazelcast.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:64)
	at com.hazelcast.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:80)
	at ------ submitted from ------.(Unknown Source)
	at com.hazelcast.client.spi.impl.ClientInvocationFuture.resolveAndThrowIfException(ClientInvocationFuture.java:96)
	at com.hazelcast.client.spi.impl.ClientInvocationFuture.resolveAndThrowIfException(ClientInvocationFuture.java:33)
	at com.hazelcast.spi.impl.AbstractInvocationFuture.get(AbstractInvocationFuture.java:155)
	at com.hazelcast.client.spi.ClientProxy.invoke(ClientProxy.java:231)
	at com.hazelcast.client.proxy.ClientMapProxy.executeOnEntries(ClientMapProxy.java:1340)
	at com.hazelcast.client.usercodedeployment.ClientUserCodeDeploymentTest.assertCodeDeploymentWorking(ClientUserCodeDeploymentTest.java:150)
	at com.hazelcast.client.usercodedeployment.ClientUserCodeDeploymentTest.testWithMultipleNodes_clientReconnectsToNewNode(ClientUserCodeDeploymentTest.java:140)
Caused by: java.lang.ClassNotFoundException: Failed to load class usercodedeployment.IncrementingEntryProcessor from other members.
	at com.hazelcast.internal.usercodedeployment.impl.ClassLocator.tryToGetClassFromRemote(ClassLocator.java:157)
	at com.hazelcast.internal.usercodedeployment.impl.ClassLocator.handleClassNotFoundException(ClassLocator.java:95)
	at com.hazelcast.internal.usercodedeployment.UserCodeDeploymentService.handleClassNotFoundException(UserCodeDeploymentService.java:89)
	at com.hazelcast.internal.usercodedeployment.UserCodeDeploymentClassLoader.loadClass(UserCodeDeploymentClassLoader.java:57)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:247)
	at com.hazelcast.nio.ClassLoaderUtil.tryLoadClass(ClassLoaderUtil.java:173)
	at com.hazelcast.nio.ClassLoaderUtil.loadClass(ClassLoaderUtil.java:147)
	at com.hazelcast.nio.IOUtil$ClassLoaderAwareObjectInputStream.resolveClass(IOUtil.java:605)
	at java.io.ObjectInputStream.readNonProxyDesc(ObjectInputStream.java:1589)
	at java.io.ObjectInputStream.readClassDesc(ObjectInputStream.java:1494)
	at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:1748)
	at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1327)
	at java.io.ObjectInputStream.readObject(ObjectInputStream.java:349)
	at com.hazelcast.internal.serialization.impl.JavaDefaultSerializers$JavaSerializer.read(JavaDefaultSerializers.java:79)
	at com.hazelcast.internal.serialization.impl.JavaDefaultSerializers$JavaSerializer.read(JavaDefaultSerializers.java:72)
	at com.hazelcast.internal.serialization.impl.StreamSerializerAdapter.read(StreamSerializerAdapter.java:48)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toObject(AbstractSerializationService.java:191)
	at com.hazelcast.client.impl.protocol.task.map.MapExecuteOnAllKeysMessageTask.createOperationFactory(MapExecuteOnAllKeysMessageTask.java:47)
	at com.hazelcast.client.impl.protocol.task.AbstractAllPartitionsMessageTask.processMessage(AbstractAllPartitionsMessageTask.java:38)
	at com.hazelcast.client.impl.protocol.task.AbstractMessageTask.initializeAndProcessMessage(AbstractMessageTask.java:130)
	at com.hazelcast.client.impl.protocol.task.AbstractMessageTask.run(AbstractMessageTask.java:110)
	at java.util.concurrent.ThreadPoolExecutor$Worker.runTask(ThreadPoolExecutor.java:895)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:918)
	at java.lang.Thread.run(Thread.java:662)
	at com.hazelcast.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:64)
	at com.hazelcast.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:80)
```

https://hazelcast-l337.ci.cloudbees.com/view/Hazelcast/job/Hazelcast-3.x-freestyle/1803/testReport/junit/com.hazelcast.client.usercodedeployment/ClientUserCodeDeploymentTest/testWithMultipleNodes_clientReconnectsToNewNode_ClassCacheMode_OFF__ProviderMode_LOCAL_CLASSES_ONLY_/

https://hazelcast-l337.ci.cloudbees.com/view/Hazelcast/job/Hazelcast-3.x-OracleJDK1.6-freestyle/402/testReport/junit/com.hazelcast.client.usercodedeployment/ClientUserCodeDeploymentTest/testWithMultipleNodes_clientReconnectsToNewNode_ClassCacheMode_ETERNAL__ProviderMode_LOCAL_AND_CACHED_CLASSES_/

--- a/hazelcast-client/src/test/java/com/hazelcast/client/usercodedeployment/ClientUserCodeDeploymentTest.java
+++ b/hazelcast-client/src/test/java/com/hazelcast/client/usercodedeployment/ClientUserCodeDeploymentTest.java
@@ -23,6 +23,8 @@ import com.hazelcast.config.Config;
 import com.hazelcast.config.UserCodeDeploymentConfig;
 import com.hazelcast.core.HazelcastInstance;
 import com.hazelcast.core.IMap;
+import com.hazelcast.core.LifecycleEvent;
+import com.hazelcast.core.LifecycleListener;
 import com.hazelcast.map.EntryProcessor;
 import com.hazelcast.test.HazelcastParametersRunnerFactory;
 import com.hazelcast.test.HazelcastTestSupport;
@@ -40,6 +42,7 @@ import usercodedeployment.IncrementingEntryProcessor;
 
 import java.util.Collection;
 import java.util.LinkedList;
+import java.util.concurrent.CountDownLatch;
 
 import static java.util.Collections.singletonList;
 import static org.junit.Assert.assertEquals;
@@ -135,8 +138,13 @@ public class ClientUserCodeDeploymentTest extends HazelcastTestSupport {
         HazelcastInstance firstInstance = factory.newHazelcastInstance(config);
         HazelcastInstance client = factory.newHazelcastClient(clientConfig);
         factory.newHazelcastInstance(config);
+
+        final CountDownLatch clientReconnectedLatch = new CountDownLatch(1);
+        client.getLifecycleService().addLifecycleListener(new ClientReconnectionListener(clientReconnectedLatch));
+
         firstInstance.getLifecycleService().shutdown();
 
+        assertOpenEventually(clientReconnectedLatch);
         assertCodeDeploymentWorking(client, new IncrementingEntryProcessor());
     }
 
@@ -170,4 +178,18 @@ public class ClientUserCodeDeploymentTest extends HazelcastTestSupport {
         assertCodeDeploymentWorking(client, new EntryProcessorWithAnonymousAndInner());
     }
 
+    private static class ClientReconnectionListener implements LifecycleListener {
+        private final CountDownLatch clientReconnectedLatch;
+
+        private ClientReconnectionListener(CountDownLatch clientReconnectedLatch) {
+            this.clientReconnectedLatch = clientReconnectedLatch;
+        }
+
+        @Override
+        public void stateChanged(LifecycleEvent event) {
+            if (event.getState() == LifecycleEvent.LifecycleState.CLIENT_CONNECTED) {
+                clientReconnectedLatch.countDown();
+            }
+        }
+    }
 }

